addons/account/models/account_reconcile_model.py

```diff
@@ -556,65 +605,145 @@ def _get_invoice_matching_query(self, st_lines, excluded_ids=None, partner_map=N
         :param excluded_ids:    Account.move.lines to exclude.
-        :param partner_map:     Dict mapping each line with new partner eventually.
         :return:                (query, params)
         '''
-        if any(m.rule_type != 'writeoff_suggestion' for m in self):
-            raise UserError(_('Programmation Error: Can\'t call _get_wo_suggestion_query() for different rules than \'writeoff_suggestion\''))
-
-        queries = []
-        all_params = []
-        for rule in self:
-            query = '''
-                SELECT
-                    %s                                  AS sequence,
-                    %s                                  AS model_id,
-                    st_line.id                          AS id
-                FROM account_bank_statement_line st_line
-                JOIN account_move st_line_move          ON st_line_move.id = st_line.move_id
-                LEFT JOIN account_journal journal       ON journal.id = st_line_move.journal_id
-                LEFT JOIN jnl_precision                 ON jnl_precision.journal_id = journal.id
-                LEFT JOIN res_company company           ON company.id = st_line_move.company_id
-                LEFT JOIN partners_table line_partner   ON line_partner.line_id = st_line.id
-                WHERE st_line.id IN %s
-            '''
-            params = [rule.sequence, rule.id, tuple(st_lines.ids)]
-
-            query, params = rule._apply_conditions(query, params)
-            queries.append(query)
-            all_params += params
-
-        full_query = self._get_with_tables(st_lines, partner_map=partner_map)
-        full_query += ' UNION ALL '.join(queries)
-        return full_query, all_params
+        self.ensure_one()
+
+        if self.rule_type != 'writeoff_suggestion':
+            raise UserError(_("Programmation Error: Can't call _get_writeoff_suggestion_query() for different rules than 'writeoff_suggestion'"))
+
+        query = '''
+            SELECT
+                %(sequence)s                        AS sequence,
+                %(model_id)s                        AS model_id,
+                st_line.id                          AS id
+            FROM account_bank_statement_line st_line
+            JOIN account_move st_line_move          ON st_line_move.id = st_line.move_id
+            LEFT JOIN account_journal journal       ON journal.id = st_line_move.journal_id
+            LEFT JOIN res_company company           ON company.id = st_line_move.company_id
+            WHERE st_line.id IN %(st_line_ids)s
+        '''
+        params = {
+            'sequence': self.sequence,
+            'model_id': self.id,
+            'st_line_ids': tuple(st_line.id for (st_line, partner) in st_lines_with_partner),
+        }
+
+        return query, params
+
+    def _get_rule_result(self, st_line, candidates, aml_ids_to_exclude, reconciled_amls_ids, partner_map):
```

Comment:
this method seems quite useless. Why not call `_get_invoice_matching_rule_result` & `_get_writeoff_suggestion_rule_result` directly in `_apply_rules`?

Reply:
To reduce apply_rules and simplify it as much as possible. A big part of what made the code unreadable before was that is was too monolithic. Here, if you need to add a new kind of rule, you basically only need to modify a few small functions, not to impact apply_rules.